django/db/models/constraints.py

```diff
@@ -132,37 +157,57 @@ def _get_condition_sql(self, model, schema_editor):
         return sql % tuple(schema_editor.quote_value(p) for p in params)
 
+    def _get_index_expressions(self, model, schema_editor):
+        if not self.expressions:
+            return None
+        index_expressions = []
+        for expression in self.expressions:
+            index_expression = IndexExpression(expression)
+            index_expression.set_wrapper_classes(schema_editor.connection)
+            index_expressions.append(index_expression)
+        return ExpressionList(*index_expressions).resolve_expression(
+            Query(model, alias_cols=False),
+        )
+
     def constraint_sql(self, model, schema_editor):
         fields = [model._meta.get_field(field_name).column for field_name in self.fields]
         include = [model._meta.get_field(field_name).column for field_name in self.include]
         condition = self._get_condition_sql(model, schema_editor)
+        expressions = self._get_index_expressions(model, schema_editor)
         return schema_editor._unique_sql(
             model, fields, self.name, condition=condition,
             deferrable=self.deferrable, include=include,
-            opclasses=self.opclasses,
+            opclasses=self.opclasses, expressions=expressions,
         )
 
     def create_sql(self, model, schema_editor):
         fields = [model._meta.get_field(field_name).column for field_name in self.fields]
         include = [model._meta.get_field(field_name).column for field_name in self.include]
         condition = self._get_condition_sql(model, schema_editor)
+        expressions = self._get_index_expressions(model, schema_editor)
         return schema_editor._create_unique_sql(
             model, fields, self.name, condition=condition,
             deferrable=self.deferrable, include=include,
-            opclasses=self.opclasses,
+            opclasses=self.opclasses, expressions=expressions,
         )
 
     def remove_sql(self, model, schema_editor):
         condition = self._get_condition_sql(model, schema_editor)
         include = [model._meta.get_field(field_name).column for field_name in self.include]
+        expressions = self._get_index_expressions(model, schema_editor)
         return schema_editor._delete_unique_sql(
             model, self.name, condition=condition, deferrable=self.deferrable,
-            include=include, opclasses=self.opclasses,
+            include=include, opclasses=self.opclasses, expressions=expressions,
         )
 
     def __repr__(self):
-        return '<%s: fields=%r name=%r%s%s%s%s>' % (
-            self.__class__.__name__, self.fields, self.name,
+        return '<%s: %sname=%r%s%s%s%s%s>' % (
+            self.__class__.__name__,
+            '' if not self.fields else 'fields=%s ' % repr(self.fields),
+            self.name,
+            '' if not self.expressions else ' expressions=%s' % ', '.join([
+                str(expression) for expression in self.expressions
+            ]),
```

Comment:
We could  keep the same format as in indexes:
```suggestion
        return '<%s: %sname=%r%s%s%s%s%s>' % (
            self.__class__.__name__,
            '' if not self.fields else "fields='%s' " % ', '.join(self.fields),
            self.name,
            '' if not self.expressions else " expressions='%s'" % ', '.join([
                str(expression) for expression in self.expressions
            ]),
```

Reply:
Yeah I've been meaning to open up a PR with that. Do you think we should change the other arguments like `condition` and `opclasses` while we're at it or do we want it in a separate commit?

Reply:
Yes a separate PR with unification sounds good.